examples/pxScene2d/src/pxShaderResource.cpp

```diff
@@ -0,0 +1,1433 @@
+      loadFrgShader = RT_OK;
+      break;
+    }
+
+    loadFrgShader = rtLoadFile( mFragmentUrl, mFragmentSrc);
+    if (loadFrgShader == RT_OK)
+      break;
+
+    if (rtIsPathAbsolute(mFragmentUrl))
+      break;
+
+    rtModuleDirs *dirs = rtModuleDirs::instance();
+
+    for (rtModuleDirs::iter it = dirs->iterator(); it.first != it.second; it.first++)
+    {
+      if (rtLoadFile(rtConcatenatePath(*it.first, mFragmentUrl.cString()).c_str(), mFragmentSrc) == RT_OK)
+      {
+        loadFrgShader = RT_OK;
+        break;
+      }
+    }
+  } while(0);
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+  rtError loadVtxShader = RT_FAIL;
+
+  do
+  {
+    if (mVertexSrc.length() != 0)
+    {
+      // We have VERTEX SHADER source already...
+      loadVtxShader = RT_OK;
+      break;
+    }
+
+    if (mVertexSrc.length() == 0)
+    {
+      //Use the Default VERTEX SHADER source...
+      loadVtxShader = RT_OK;
+      break;
+    }
+
+    loadVtxShader = rtLoadFile(mVertexUrl, mVertexSrc);
+    if (loadVtxShader == RT_OK)
+      break;
+
+    if (rtIsPathAbsolute(mVertexUrl))
+      break;
+
+    rtModuleDirs *dirs = rtModuleDirs::instance();
+
+    for (rtModuleDirs::iter it = dirs->iterator(); it.first != it.second; it.first++)
+    {
+      if (rtLoadFile(rtConcatenatePath(*it.first, mVertexUrl.cString()).c_str(), mVertexSrc) == RT_OK)
+      {
+        loadVtxShader = RT_OK;
+        break;
+      }
+    }
+  } while(0);
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+  // CREATE SHADER PROGRAMS
+
+  if (loadFrgShader == RT_OK && loadVtxShader == RT_OK)
+  {
+    setupResource();
+  }
+  else
+  {
+    loadFrgShader = RT_RESOURCE_NOT_FOUND;
+    rtLogError("Could not load FRAGMENT shader file %s.", mFragmentUrl.cString());
+  }
+  if ( loadFrgShader != RT_OK)
+  {
+    rtLogWarn("shader load failed"); // TODO: why?
+    if (loadFrgShader == RT_RESOURCE_NOT_FOUND)
+    {
+      setLoadStatus("statusCode",PX_RESOURCE_STATUS_FILE_NOT_FOUND);
+    }
+    else
+    {
+      setLoadStatus("statusCode", PX_RESOURCE_STATUS_DECODE_FAILURE);
+    }
+
+//    // Since this object can be released before we get a async completion
+//    // We need to maintain this object's lifetime
+//    // TODO review overall flow and organization
+//    AddRef();
+
+    if (gUIThreadQueue)
+    {
+      gUIThreadQueue->addTask(onDownloadCanceledUI, this, (void*)"reject");
+    }
+    //mTexture->notifyListeners( mTexture, RT_FAIL, errorCode);
+  }
+  else
+  {
+    mFragmentSrc.term(); // Dump the source data...
+    mVertexSrc.term();   // Dump the source data...
+
+    setLoadStatus("statusCode",0);
+    // Since this object can be released before we get a async completion
+    // We need to maintain this object's lifetime
+    // TODO review overall flow and organization
+    AddRef();
```

Comment:
Just curious, why do we need to call `AddRef();` twice inside of `loadResourceFromFile()`?  I understand a call to `AddRef()` before calling `gUIThreadQueue->addTask()`, but `AddRef()` is always called on this object at the beginning of this method as well.  Why the need for two calls?

Reply:
Removed